modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java

```diff
@@ -0,0 +1,321 @@
+    private volatile TimeValue pollInterval;
+    private SchedulerEngine.Job scheduledJob;
+    private final SetOnce<SchedulerEngine> scheduler = new SetOnce<>();
+    // we use this rollover supplier to facilitate testing until we'll be able to read the
+    // rollover configuration from a cluster setting
+    private Function<String, RolloverRequest> defaultRolloverRequestSupplier;
+
+    public DataLifecycleService(
+        Settings settings,
+        Client client,
+        ClusterService clusterService,
+        Clock clock,
+        ThreadPool threadPool,
+        LongSupplier nowSupplier
+    ) {
+        this.settings = settings;
+        this.client = client;
+        this.clusterService = clusterService;
+        this.clock = clock;
+        this.transportActionsDeduplicator = new ResultDeduplicator<>(threadPool.getThreadContext());
+        this.nowSupplier = nowSupplier;
+        this.scheduledJob = null;
+        this.pollInterval = DLM_POLL_INTERVAL_SETTING.get(settings);
+        this.defaultRolloverRequestSupplier = this::getDefaultRolloverRequest;
+    }
+
+    /**
+     * Initializer method to avoid the publication of a self reference in the constructor.
+     */
+    public void init() {
+        clusterService.addListener(this);
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(DLM_POLL_INTERVAL_SETTING, this::updatePollInterval);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        // wait for the cluster state to be recovered
+        if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
+            return;
+        }
+
+        final boolean prevIsMaster = this.isMaster;
+        if (prevIsMaster != event.localNodeMaster()) {
+            this.isMaster = event.localNodeMaster();
+            if (this.isMaster) {
+                // we weren't the master, and now we are
+                onBecomeMaster();
+            } else {
+                // we were the master, and now we aren't
+                cancelJob();
+            }
+        }
+    }
+
+    @Override
+    public void close() {
+        SchedulerEngine engine = scheduler.get();
+        if (engine != null) {
+            engine.stop();
+        }
+    }
+
+    @Override
+    public void triggered(SchedulerEngine.Event event) {
+        if (event.getJobName().equals(DATA_LIFECYCLE_JOB_NAME)) {
+            logger.trace("DLM job triggered: " + event.getJobName() + ", " + event.getScheduledTime() + ", " + event.getTriggeredTime());
+            run(clusterService.state());
+        }
+    }
+
+    /**
+     * Iterates over the DLM managed data streams and executes the needed operations
+     * to satisfy the configured {@link org.elasticsearch.cluster.metadata.DataLifecycle}.
+     */
+    // default visibility for testing purposes
+    void run(ClusterState state) {
+        for (DataStream dataStream : state.metadata().dataStreams().values()) {
```

Comment:
We need error handling around these, any exception right now would short-circuit and prevent rollover or retention from being invoked in a later data stream. Ideally I think we should have error handling for each piece individually also (to be defensive), so that a failure somewhere in our code doesn't interrupt flow.

Alternatively, if we factored these pieces into separate methods also, then this could look like:

```java
for (DataStream dataStream : state.metadata().dataStreams().values()) {
    if (dataStream.getLifecycle() == null) {
        continue;
    }
    
    try {
        maybeInvokeRollover(state, dataStream);
    } catch (Exception e) {
        logger.error("failed to roll over etc etc", e);
    }

    try {
         maybeInvokeRetention(state, dataStream);
    } catch (Exception e) {
        logger.error("failed to do all the retentiony stuff", e);
    }
}
```

Reply:
Done